CelesteNet.Client/CelesteNetClientTCPUDPConnection.cs

```diff
@@ -0,0 +1,337 @@
+using Celeste.Mod.CelesteNet.DataTypes;
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+
+namespace Celeste.Mod.CelesteNet.Client {
+    public class CelesteNetClientTCPUDPConnection : CelesteNetTCPUDPConnection {
+
+        public const int TCPBufferSize = 16384;
+        public const int UDPBufferSize = 16384;
+        public const int UDPEstablishDelay = 2;
+        private static readonly byte[] UDPHolePunchMessage = new byte[] { 42 };
+
+        private BufferedSocketStream tcpStream;
+        private Socket udpRecvSocket, udpSendSocket;
+        private BlockingCollection<DataType> tcpSendQueue, udpSendQueue;
+
+        private CancellationTokenSource tokenSrc;
+        private Thread tcpRecvThread, udpRecvThread, tcpSendThread, udpSendThread;
+
+        public CelesteNetClientTCPUDPConnection(DataContext data, uint token, Settings settings, Socket tcpSock) : base(data, token, settings, tcpSock, FlushTCPQueue, FlushUDPQueue) {
+            // Initialize networking
+            tcpSock.Blocking = false;
+            tcpStream = new BufferedSocketStream(TCPBufferSize) { Socket = tcpSock };
+            IPAddress serverAddr = ((IPEndPoint) tcpSock.RemoteEndPoint).Address;
+
+            udpRecvSocket = new Socket(SocketType.Dgram, ProtocolType.Udp);
+            udpRecvSocket.EnableEndpointReuse();
+            udpRecvSocket.Connect(serverAddr, settings.UDPSendPort);
+
+            udpSendSocket = new Socket(SocketType.Dgram, ProtocolType.Udp);
+            udpSendSocket.EnableEndpointReuse();
+            udpSendSocket.Bind(udpRecvSocket.LocalEndPoint);
+            udpSendSocket.Connect(serverAddr, settings.UDPReceivePort);
+
+            OnUDPDeath += (_, _) => {
+                if (CelesteNetClientModule.Instance?.Context?.Status != null)
+                    CelesteNetClientModule.Instance.Context.Status.Set(UseUDP ? "UDP connection died" : "Switching to TCP only", 3);
+            };
```

Comment:
A mix of both. It was used for debugging, but it might be useful to let the player know when the UDP connection dies (which might have an impact on performance), so that they can switch to TCP-only manually if they notice they have an unstable UDP connection, which skips the initial lag when the client initially tries it.